mmdet/models/dense_heads/yolact_head.py

```diff
@@ -0,0 +1,680 @@
+                cur_segm_pred,
+                segm_targets,
+                avg_factor=num_imgs * mask_h * mask_w)
+            loss_segm.append(loss)
+        return dict(loss_segm=loss_segm)
+
+
+@HEADS.register_module()
+class YolactProtonet(nn.Module):
+    """YOLACT mask head used in https://arxiv.org/abs/1904.02689.
+
+    This head outputs the mask prototypes for YOLACT.
+    """
+
+    def __init__(
+        self,
+        in_channels=256,
+        num_protos=32,
+        num_classes=80,
+        loss_mask_weight=1.0,
+        max_masks_to_train=100,
+    ):
+        super(YolactProtonet, self).__init__()
+        config = [(256, 3, {'padding': 1})] * 3 + \
+            [(None, -2, {}), (256, 3, {'padding': 1})] \
+            + [(num_protos, 1, {})]
+        self.protonet = self.make_net(in_channels, config)
+        self.loss_mask_weight = loss_mask_weight
+        self.num_protos = num_protos
+        self.num_classes = num_classes
+        self.max_masks_to_train = max_masks_to_train
+        self.fp16_enabled = False
+
+    def make_net(self, in_channels, config, include_last_relu=True):
+        """A helper function to take a config setting and turn it into a
+        network."""
+
+        def make_layer(layer_cfg):
+            nonlocal in_channels
+
+            # Possible patterns:
+            # ( 256, 3, {}) -> conv
+            # ( 256,-2, {}) -> deconv
+            # (None,-2, {}) -> bilinear interpolate
+
+            num_channels = layer_cfg[0]
+            kernel_size = layer_cfg[1]
+
+            if kernel_size > 0:
+                layer = nn.Conv2d(in_channels, num_channels, kernel_size,
+                                  **layer_cfg[2])
+            else:
+                if num_channels is None:
+                    layer = InterpolateModule(
+                        scale_factor=-kernel_size,
+                        mode='bilinear',
+                        align_corners=False,
+                        **layer_cfg[2])
+                else:
+                    layer = nn.ConvTranspose2d(in_channels, num_channels,
+                                               -kernel_size, **layer_cfg[2])
+
+            in_channels = num_channels if num_channels is not None \
+                else in_channels
+
+            return [layer, nn.ReLU(inplace=True)]
+
+        # Use sum to concat together all the component layer lists
+        net = sum([make_layer(x) for x in config], [])
+        if not include_last_relu:
+            net = net[:-1]
+
+        return nn.Sequential(*(net))
+
+    def init_weights(self):
+        for m in self.protonet:
+            if isinstance(m, nn.Conv2d):
+                xavier_init(m, distribution='uniform')
+
+    def forward(self, x, coeff_pred, bboxes, img_meta, sampling_results=None):
+        prototypes = self.protonet(x)
+        prototypes = prototypes.permute(0, 2, 3, 1).contiguous()
+
+        num_imgs = x.size(0)
+        # Training state
+        if sampling_results is not None:
```

Comment:
If I am not wrong, here we use `sampling_results is None` to decide if the model is training. But we can also use `self.training` to decide if is training or not. It will be more straightforward.